File: IT/2/Multithreading/Esercizi/Java/Esercizio10.java
/* file Esercizio10.java */

import java.util.Arrays;

class Target implements Runnable {
	private int [] arr;
	private int start;
	private int end;

	public Target(int [] arr, int start, int end) {
		this.arr = arr;
		this.start = start;
		this.end = end;
	}

	@Override
	public void run() {
		for (int i = start; i < end; i++) {
			arr[i] = i + 1;
		}
	}
}

public class Esercizio10 {
	public static void main(String [] args) {
		int [] arr = new int[9];
		System.out.println(Arrays.toString(arr));

		try {
			Thread t1 = new Thread(new Target(arr, 0, 3));
			Thread t2 = new Thread(new Target(arr, 3, 6));
			Thread t3 = new Thread(new Target(arr, 6, 9));

			t1.start();
			t2.start();
			t3.start();

			t1.join();
			t2.join();
			t3.join();

			System.out.println(Arrays.toString(arr));
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio10.py
# file esercizio10.py

from __future__ import print_function
from multiprocessing import Process, RawArray
from sys import stdout
from os import linesep

def target(arr, start, end):
	for i in range(start, end):
		arr[i] = i+1

if __name__ == "__main__":
	arr = RawArray('i', [0]*9)

	for i in arr:
		print(i, end=" ")

	stdout.write(linesep)

	procs = [Process(target=target, args=(arr, i, i+3,)) for i in range(0, 9, 3)]

	for p in procs:
		p.start()

	for p in procs:
		p.join()

	for i in arr:
		print(i, end=" ")

	stdout.write(linesep)

